examples/step-77/step-77.cc

```diff
@@ -372,19 +365,12 @@ namespace Step77
           }
 
         cell->get_dof_indices(local_dof_indices);
-        for (unsigned int i = 0; i < dofs_per_cell; ++i)
-          residual(local_dof_indices[i]) += cell_residual(i);
+        zero_constraints.distribute_local_to_global(cell_residual,
+                                                    local_dof_indices,
+                                                    residual);
       }
 
-    hanging_node_constraints.condense(residual);
-
-    for (const types::global_dof_index i :
-         DoFTools::extract_boundary_dofs(dof_handler))
-      residual(i) = 0;
-
-    for (const types::global_dof_index i :
-         DoFTools::extract_hanging_node_dofs(dof_handler))
-      residual(i) = 0;
+    zero_constraints.set_zero(residual);
```

Comment:
Is this actually necessary? Same for the corresponding call below.

Reply:
Right. The call to `zero_constraints.set_zero()` happened right after this:
```
 zero_constraints.distribute_local_to_global(cell_residual,
                                                    local_dof_indices,
                                                    residual);
```
But this call should have set the right hand side entries corresponding to constrained DoFs to zero (or, better: Should have added zeros to these entries). I don't think you need to zero them out explicitly.

As for `distribute()` after refinement: I think I've noticed this before, and we probably do it in all tutorials that use `SolutionTransfer`. That class does not enforce constraints because it works cell-by-cell and does not have its own constraints object at the time we do the interpolation to the new mesh. It makes sense to keep the `distribute()` in. You'd end up with an unnecessarily large residual otherwise, and in particular with a solution vector that does not satisfy constraints.